### PR TITLE
CFE-2893: Added a new option --skip-db-check to cf-agent and cf-execd

### PR DESCRIPTION
**Pull request by olehermanse**
This option allows you to enable/disable database (LMDB) consistency checks. Initially it is disabled by default, but this will likely change.

```
root@dev core $ /var/cfengine/bin/cf-agent --skip-db-check=no
root@dev core $ /var/cfengine/bin/cf-agent --skip-db-check=no --log-level info
    info: No filenames specified, defaulting to .lmdb files in /var/cfengine/state/
    info: Status of '/var/cfengine/state/cf_state.lmdb': OK
    info: Status of '/var/cfengine/state/history.lmdb': OK
    info: Status of '/var/cfengine/state/packages_installed_apt_get.lmdb': OK
    info: Status of '/var/cfengine/state/nova_measures.lmdb': OK
    info: Status of '/var/cfengine/state/nova_static.lmdb': OK
    info: Status of '/var/cfengine/state/packages_updates_apt_get.lmdb': OK
    info: Status of '/var/cfengine/state/cf_lastseen.lmdb': OK
    info: Status of '/var/cfengine/state/cf_observations.lmdb': OK
    info: Status of '/var/cfengine/state/cf_lock.lmdb': OK
    info: Status of '/var/cfengine/state/packages_installed_$(package_module_knowledge.platform_default).lmdb': OK
    info: All 10 databases healthy
    info: No corruption - nothing to do
root@dev core $ /var/cfengine/bin/cf-execd --skip-db-check=no --log-level info
    info: No filenames specified, defaulting to .lmdb files in /var/cfengine/state/
    info: Status of '/var/cfengine/state/cf_state.lmdb': OK
    info: Status of '/var/cfengine/state/history.lmdb': OK
    info: Status of '/var/cfengine/state/packages_installed_apt_get.lmdb': OK
    info: Status of '/var/cfengine/state/nova_measures.lmdb': OK
    info: Status of '/var/cfengine/state/nova_static.lmdb': OK
    info: Status of '/var/cfengine/state/packages_updates_apt_get.lmdb': OK
    info: Status of '/var/cfengine/state/cf_lastseen.lmdb': OK
    info: Status of '/var/cfengine/state/cf_observations.lmdb': OK
    info: Status of '/var/cfengine/state/cf_lock.lmdb': OK
    info: Status of '/var/cfengine/state/packages_installed_$(package_module_knowledge.platform_default).lmdb': OK
    info: All 10 databases healthy
    info: No corruption - nothing to do
    info: cf-execd starting Wed Jan  9 15:01:55 2019
root@dev core $
```

### CODE REVIEW

**Comment by olehermanse** [comment]
@cf-bottom jenkins with exotics.

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1920/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1920/

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1922/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1922/

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1929/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1929/

**Comment by cf-bottom** [comment]
Sure, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1935/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1935/

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1936/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1936/

File: cf-check/repair.c [hunk 1]
```diff
@@ -116,8 +116,30 @@ int repair_files(Seq *files)
...
+    {
+        return 1;
+    }
     const int ret = repair_files(files);
     SeqDestroy(files);
+
+    if (ret != 0)
+    {
+        Log(LOG_LEVEL_ERR, "Something went wrong during database repair\n");
+        Log(LOG_LEVEL_ERR, "Try running `cf-check repair` manually\n");
```
Comment (vpodzime) [inline cf-check/repair.c]: The `\n` should not be there for `Log()` calls.

File: libutils/file_lib.h [hunk 2]
```diff
@@ -88,6 +100,7 @@ char *MapName(char *s);
 char *MapNameCopy(const char *s);
 char *MapNameForward(char *s);
 
+Seq *ls(const char *dir, const char *extension);
```
Comment (vpodzime) [inline libutils/file_lib.h]: I think `ListDir()` or something like that would be a better name now that the function is possibly used more widely.

File: libutils/path.c [hunk 3]
```diff
@@ -0,0 +1,34 @@
+#include <path.h>
+#include <string_lib.h>
+
+const char *Path_Basename(const char *path)
```
Comment (vpodzime) [inline libutils/path.c]: Might be useful to mention why the `basename()` function from *glibc* is not good.

File: libutils/string_lib.c [hunk 4]
```diff
@@ -1174,6 +1174,7 @@ char *TrimWhitespace(char *s)
     return s;
 }
 
+
```
Comment (vpodzime) [inline libutils/string_lib.c]: Was this change intentional?

File: libutils/path.c (hunk 3 above)
Comment (olehermanse) [reply libutils/path.c]: It's in the header.

File: libutils/file_lib.h (hunk 2 above)
Comment (olehermanse) [reply libutils/file_lib.h]: Fixed.

File: libutils/string_lib.c (hunk 4 above)
Comment (olehermanse) [reply libutils/string_lib.c]: Nope, thanks.

**Comment by olehermanse** [comment]
@cf-bottom again in jenkins with exotics.

**Comment by cf-bottom** [comment]
Alright, I triggered a build:

[Build Status](https://ci.cfengine.com//job/pr-pipeline/1942/)

(with exotics)

https://ci.cfengine.com/job/pr-pipeline/1942/

File: Makefile.am [hunk 5]
```diff
@@ -32,8 +32,8 @@ SUBDIRS = libcompat \
 	libcfnet \
 	libenv \
 	libpromises \
-	cf-agent \
 	cf-check \
+	cf-agent \
```
Comment (craigcomstock) [inline Makefile.am]: Is this change a dependency sort of change? Like cf-check must be built before cf-agent?

File: cf-agent/cf-agent.c [hunk 6]
```diff
@@ -102,6 +103,7 @@ extern int PR_NOTKEPT;
 static bool ALLCLASSESREPORT = false; /* GLOBAL_P */
 static bool ALWAYS_VALIDATE = false; /* GLOBAL_P */
 static bool CFPARANOID = false; /* GLOBAL_P */
+static bool PERFORM_DB_CHECK = false;
```
Comment (craigcomstock) [inline cf-agent/cf-agent.c]: of what significance is the comment /* GLOBAL_P */ ?

File: cf-check/backup.c [hunk 7]
```diff
@@ -30,28 +32,38 @@ const char *create_backup_dir()
...
-    const int n = snprintf(backup_dir, PATH_MAX, "%s%jd/", backup_root, (intmax_t) ts);
+    const int n =
+        snprintf(backup_dir, PATH_MAX, "%s%jd/", backup_root, (intmax_t)ts);
     if (n >= PATH_MAX)
     {
-        printf("Backup path too long: %jd/%jd\n", (intmax_t) n, (intmax_t) PATH_MAX);
+        Log(LOG_LEVEL_ERR,
+            "Backup path too long: %jd/%jd",
+            (intmax_t)n,
+            (intmax_t)PATH_MAX);
```
Comment (craigcomstock) [inline cf-check/backup.c]: at first I was confused by the use of intmax_t. That is the largest signed integer type but wasn't sure why you used it here. So please give me a hint and educate me. :)

File: libutils/file_lib.c [hunk 8]
```diff
@@ -67,6 +68,112 @@ Writer *FileRead(const char *filename, size_t max_size, bool *truncated)
...
+    }
+
+    size_t bytes_in = 0;
+    size_t bytes_out = 0;
+    bool ret = true;
+    do
+    {
+#define BUFSIZE 1024
+        char buf[BUFSIZE] = {0};
+
```
Comment (craigcomstock) [inline libutils/file_lib.c]: Not that performance matters too much but is there a reason to initialize the buf variable inside the block instead of outside of it?

File: cf-check/backup.c (hunk 7 above)
Comment (olehermanse) [reply cf-check/backup.c]: Note that this change is only a formatting change, so I shouldn't change anything here.

When you don't have an appropriate `%` specifier, the best approach is to cast to `uintmax_t` or `intmax_t` and use `%ju` or `%jd`. In this case it was probably because I don't know what type `PATH_MAX` will be on all systems. (`n` is just an `int` and could be printed using `%d`).

File: Makefile.am (hunk 5 above)
Comment (olehermanse) [reply Makefile.am]: It was, `libcf-check.la` was compiled inside `cf-check` and then used for `cf-agent`. However, since I switched to including `cf-check` in `libpromises` instead, it's no longer necessary.

File: cf-agent/cf-agent.c (hunk 6 above)
Comment (olehermanse) [reply cf-agent/cf-agent.c]: Don't know.

File: libutils/file_lib.c (hunk 8 above)
Comment (olehermanse) [reply libutils/file_lib.c]: Nothing too strong. We tend to prefer to have the variables be as local as possible, for readability / maintainability. It is also safer to make sure it's never dirty. Doing it outside the loop for performance would be valid as well, but I don't think performance is that critical here. IO performance (fwrite and fread) probably makes this difference completely negligible.
